pretyping/tacred.ml

```diff
@@ -827,10 +744,102 @@ and whd_simpl_stack env sigma =
   in
   redrec
 
+and reduce_fix env sigma fix stack =
+  match fix_recarg fix stack with
+    | None -> NotReducible
+    | Some (recargnum,recarg) ->
+        let (recarg'hd,_ as recarg') = whd_construct_stack env sigma recarg in
+        let stack' = List.assign stack recargnum (applist recarg') in
+        (match EConstr.kind sigma recarg'hd with
+           | Construct _ -> Reduced (contract_fix sigma fix, stack')
+           | _ -> NotReducible)
+
+and reduce_fix_use_function env sigma f fix stack =
+  match fix_recarg fix stack with
+    | None -> NotReducible
+    | Some (recargnum,recarg) ->
+        let (recarg'hd,_ as recarg') =
+          if EConstr.isRel sigma recarg then
+            (* The recarg cannot be a local def, no worry about the right env *)
+            (recarg, [])
+          else
+            whd_construct_stack env sigma recarg in
+        let stack' = List.assign stack recargnum (applist recarg') in
+        (match EConstr.kind sigma recarg'hd with
+           | Construct _ ->
+              let (names, (nbfix, lv, n)), u, largs = f in
+              let lu = List.firstn n largs in
+              let p = List.length lv in
+              let lyi = List.map fst lv in
+              let la =
+                List.map_i (fun q aq ->
+                  (* k from the comment is q+1 *)
+                  try mkRel (p+1-(List.index Int.equal (n-q) lyi))
+                  with Not_found -> Vars.lift p aq)
+                  0 lu
+              in
+              let f i = match names.(i) with
+              | None -> None
+              | Some (minargs,ref) ->
+                  let body = applist (mkEvalRef ref u, la) in
+                  let g =
+                    List.fold_left_i (fun q (* j = n+1-q *) c (ij,tij) ->
+                      let subst = List.map (Vars.lift (-q)) (List.firstn (n-ij) la) in
+                      let tij' = Vars.substl (List.rev subst) tij in
+                      let x = make_annot x Sorts.Relevant in  (* TODO relevance *)
+                      mkLambda (x,tij',c)) 1 body (List.rev lv)
+                  in Some (minargs,g)
+              in
+               Reduced (contract_fix_use_function env sigma f fix,stack')
+           | _ -> NotReducible)
+
+and reduce_proj env sigma c =
+  let rec redrec s =
+    match EConstr.kind sigma s with
+    | Proj (proj, c) ->
+      let c' = try redrec c with Redelimination -> c in
+      let constr, cargs =  whd_construct_stack env sigma c' in
+        (match EConstr.kind sigma constr with
+        | Construct _ ->
+          let proj_narg = Projection.npars proj + Projection.arg proj in
+          List.nth cargs proj_narg
+        | _ -> raise Redelimination)
+    | Case (n,p,iv,c,brs) ->
+      let c' = redrec c in
+      let p = (n,p,iv,c',brs) in
+        (try special_red_case env sigma p
+         with Redelimination -> mkCase p)
+    | _ -> raise Redelimination
+  in redrec c
+
+and special_red_case env sigma (ci, p, iv, c, lf) =
```

Comment:
I'd love a comment, since "special" is not very clear

Reply:
No idea why it's called that way, this PR is just removing the higher-order hoop and inlining the result.

Reply:
I've been staring at the code for a while and I still don't get what this does (of special).
I did blame it and it appeared 21 years ago in a commit by @backtracking . I'll buy you a beer if you recall why you called this function this way, back then ;-)

Reply:
I don't drink alcohol :-)
And I don't think I would be able to tell you what it does, anyway. I'm sorry.
It happens that I refactored a lot of code while making V7, in order for the code to be more consistent (in style, at least).
So it is likely I'm blamed for these lines only because of that.

Reply:
Thanks! @ppedrot I guess you can just sneak in my bounty as a comment ;-)